rfcs/0107-version-normalization.md

```diff
@@ -0,0 +1,211 @@
+therefore, we should strive to follow it, neither circumventing nor ignoring it.
+
+Further, the `version` attribute should be crafted to satisfy the expected
+upgrading semantics stated in `nix-env` manpage and implemented by
+`builtins.compareVersions` -- even when the raw version of the original program
+does not meet this expectation.
+
+This document describes a format suitable to fix these issues, while keeping it
+understandable and striving for simplicity.
+
+# Detailed design
+[design]: #detailed-design
+
+## Terminology
+
+_Disclaimer_: whereas some of the terms enumerated below are borrowed from
+[pkgsrc guide](https://www.netbsd.org/docs/pkgsrc/) plus a bit of terminology
+employed by git, this document aims to be general, not being overly attached to
+them.
+
+- Program denotes the piece of software to be fetched and processed via Nixpkgs.
+  - This term makes no distinction about the immediate or intended uses of the
+    program; it can range from non-interactive programs to full GUI
+    applications, even including filesets intended as input to other programs
+    (such as firmwares and fonts).
+- Team denotes the maintainers of the program.
+  - This term makes no distinction among the various models of organization of a
+    team; it ranges from a solitary programmer to a business company - and
+    everything inbetween.
+- Source denotes the origin of the program.
+  - This term makes no distinction among precompiled, binary-only or high-level
+    code distributions.
+- Snapshot denotes the source of the program taken at a point of the time.
+  - Labeled snapshot denotes any snapshot explicitly labeled by its team.
+  - Unlabeled snapshot denotes any snapshot not explicitly labeled by its team.
+- Release denotes any distributed snapshot, as defined by its team.
+- Branch denotes a logical sequence of snapshots, as identified by the program's
+  team;
+  - Usually these branches are denoted by names such as `stable`, `master`,
+    `unstable`, `trunk`, `experimental`, `staging`, `X.Y` (where `X` and
+    `Y` are numbers) etc.
+
+## Design
+
+- For a labeled snapshot:
+  - `version` should be constituted of the version of the snapshot, as defined
+    by the program team, without any alphabetical characters (e.g. "v", "rel")
+    prepending it.
+    - Alphabetical characters following the first numerical character of version
+      (as defined above) should be maintained, except optionally those clearly
+      used as separators, in which case they are replaced by dots (emulating a
+      typical dot-separated version).
+
+- For an unlabeled snapshot:
+  - `version` should be constituted of a concatenation of the elements below in
+    this order:
+      - the version of the latest previous labeled snapshot (on the same branch,
+        when applicable), according to the rules defined before for labeled
+        snapshots;
+        - If the project never released a labeled snapshot, `0.pre` should be
```

Comment:
Because `0.0.0` has a small chance of being captured in the wild. I did not remember to test `0pre`, tho.

The idea is that unlabeled snapshots of a never-released program are the smallest possible; `0.pre` looks like a reasonable choice.

Reply:
Hmmm. Maybe someone else who is familiar how nix works internal can suggest what a good value would be because I don't feel comfortable/informed enough to make a good decision.

Reply:
Basically, pre is a -∞ component, and `0pre` and `0.pre` are equal w.r.t. `builtins.compareVersions`. `0.pre` precedes `0.0.pre`, and is preceded by `0.pre.pre`, but the latter looks a bit ridiculuous.